xarray/core/indexing.py

```diff
@@ -298,17 +299,29 @@ def slice_slice(old_slice: slice, applied_slice: slice, size: int) -> slice:
     return slice(start, stop, step)
 
 
+def slice_slice_by_array(old_slice, array, size):
+    normalized = normalize_slice(old_slice, size)
+    new_indexer = array * normalized.step + normalized.start
```

Comment:
I don't understand this. Can you add some type hints here please

Reply:
OK I think I get it. A couple of doctests would be useful, particularly one with `normalized.step > 1`

Reply:
Yeah I think those are best inlined as doctests here. They serve as docs too just like `_normalize_slice` and others in this module.